gpt_bpe.go

```diff
@@ -1140,31 +1243,264 @@ func (encoder *GPTEncoder) makeWordSplitter(
 				}
 			}
 
+			// If we have no runes, we are done
 			if len(runeAccumulator) == 0 {
-				flushBatch()
+				appendBatch(nil, true)
 				wordCallback(nil)
 				break
 			}
 
-			var line string
-			if specialToken {
-				line = string(
+			// For limiting number size
+			curNumberSize := 0
+
+			var whitespaceBuffer []rune
+			previousRuneType := -1
+			previousRuneGroupSize := 0
+
+			// Apply replacements and normalization
+			var runeLine []rune
+			if specialToken && candidateNode != nil {
+				runeLine =
 					runeAccumulator[:len(runeAccumulator)-len(
 						candidateNode.runes,
-					)],
-				)
+					)]
 			} else {
-				line = string(runeAccumulator)
+				runeLine = runeAccumulator
+			}
+			if len(encoder.replacements) > 0 {
+				runeLine = replaceRunes(
+					runeLine, encoder.replacements,
+				)
+			}
+
+			if encoder.Normalizer != nil {
+				if encoder.normalizerStringMap != nil && len(encoder.normalizerStringMap) > 0 {
+					runeLine = replaceRunes(
+						runeLine, encoder.normalizerStringMap,
+					)
+				}
+			}
+			runeAccumulator = runeLine
+			// If we don't recognize the regex, we default to using the regex package
+			if useDefaultRegexPackage && encoder.pattern.String() != "" {
```

Comment:
This is fine as is, but you might consider doing something like this, and then not having to have a comment explaining it:

```suggestion
                         regexNotRecognized := useDefaultRegexPackage && encoder.pattern.String() != ""
			if regexNotRecognized {
```